File: php/src/Webhooks.php
<?php

declare(strict_types=1);

namespace Kalourmade\Sailup;

use Kalourmade\Sailup\Exceptions\InvalidSignatureError;

final class Webhooks
{
    /**
     * @param array<string,string> $headers
     * @return array<string,mixed>
     */
    public static function verify(string $rawBody, array $headers, string $secret): array
    {
        $normalized = array_change_key_case($headers, CASE_LOWER);
        $signature = $normalized['x-sailup-signature'] ?? null;
        $timestamp = $normalized['x-sailup-timestamp'] ?? null;

        if ($signature === null || $timestamp === null) {
            throw new InvalidSignatureError('Missing signature headers');
        }

        $expected = hash_hmac('sha256', "{$timestamp}.{$rawBody}", $secret);

        if (!hash_equals($expected, $signature)) {
            throw new InvalidSignatureError('Signature mismatch');
        }

        return json_decode($rawBody, true, flags: JSON_THROW_ON_ERROR);
    }
}

File: python/src/kalourmade_sailup/webhooks.py
import hashlib
import hmac
import json
from typing import Mapping

from .exceptions import InvalidSignatureError


class Webhooks:
    @staticmethod
    def verify(raw_body: str, headers: Mapping[str, str], secret: str) -> dict:
        headers_lower = {k.lower(): v for k, v in headers.items()}
        signature = headers_lower.get("x-sailup-signature")
        timestamp = headers_lower.get("x-sailup-timestamp")

        if not signature or not timestamp:
            raise InvalidSignatureError("Missing signature headers")

        expected = hmac.new(
            secret.encode(), f"{timestamp}.{raw_body}".encode(), hashlib.sha256
        ).hexdigest()

        if not hmac.compare_digest(expected, signature):
            raise InvalidSignatureError("Signature mismatch")

        return json.loads(raw_body)
